Implement type name resolution for ILLink analyzer

sbomer opened the pull request:
Fixes https://github.com/dotnet/runtime/issues/95118

sbomer committed:
Implement type name resolution for ILLink analyzer

sbomer committed:
Don't change partial method signature

sbomer committed:
Don't keep members of pointer or byref element types

sbomer committed:
Cleanup

sbomer committed:
Revert "Don't keep members of pointer or byref element types"

This reverts commit 0e0f51d48abc92d0df9dba321c58d6ef6de703de.

sbomer committed:
Update expected behavior, add issue link

jtschuster commented on src/tools/illink/src/ILLink.RoslynAnalyzer/TrimAnalysis/RequireDynamicallyAccessedMembersAction.cs:
```diff
@@ -1,40 +1,55 @@
 	internal partial struct RequireDynamicallyAccessedMembersAction
 	{
+		readonly Compilation _compilation;
+		readonly Location _location;
+		readonly Action<Diagnostic>? _reportDiagnostic;
```
Do we need these in addition to the `_diagnosticContext`? I don't see anywhere where they would have different values

jtschuster commented on src/tools/illink/src/ILLink.RoslynAnalyzer/TrimAnalysis/TypeNameResolver.cs:
```diff
@@ -0,0 +1,150 @@
+
+namespace ILLink.Shared.TrimAnalysis
+{
+	internal partial struct TypeNameResolver
```
nit: Is there a reason this is partial?

jtschuster commented on src/tools/illink/src/ILLink.RoslynAnalyzer/TrimAnalysis/RequireDynamicallyAccessedMembersAction.cs:
```diff
@@ -1,40 +1,55 @@
 {
 	internal partial struct RequireDynamicallyAccessedMembersAction
 	{
+		readonly Compilation _compilation;
```
nit: It looks like this `Compilation` is only ever passed down to `ReflectionAccessAnalyzer.TryResolveTypeNameAndMark()` to create the same `TypeNameResolver`. Maybe we could just have a `TypeNameResolver` field instead? And we could then call `TryResolveTypeNameAndMark` directly on that field and remove `ReflectionAccessAnalyzer.TryResolveTypeNameAndMark()`.

jtschuster commented on src/tools/illink/test/Mono.Linker.Tests.Cases/DataFlow/AssemblyQualifiedNameDataflow.cs:
```diff
@@ -58,7 +58,7 @@ static void TestConstructors ()
 		}
 
-		[ExpectedWarning ("IL2122", "Type 'System.Invalid.TypeName' is not assembly qualified. " + "Type name strings used for dynamically accessing a type should be assembly qualified.", Tool.Trimmer | Tool.NativeAot, "https://github.com/dotnet/runtime/issues/95118")]
+		[ExpectedWarning ("IL2122", "Type 'System.Invalid.TypeName' is not assembly qualified. " + "Type name strings used for dynamically accessing a type should be assembly qualified.")]
```
nit:

```suggestion
		[ExpectedWarning ("IL2122", "Type 'System.Invalid.TypeName' is not assembly qualified. Type name strings used for dynamically accessing a type should be assembly qualified.")]
```

jtschuster reviewed (approved):
A couple style/organization nits, but LGTM otherwise, thank you!

sbomer committed:
PR feedback

- Remove unused using
- Simplify string concatenation
- Remove unnecessary partial
- Pass around TypeNameResolver instead of Compilation

sbomer committed:
Make it a field

jakobbotsch commented:
The repo no longer builds locally for me after this PR. I get
```
...
  LibraryImportGenerator -> C:\dev\dotnet\runtime4\artifacts\bin\LibraryImportGenerator\Release\netstandard2.0\Microsoft.Interop.LibraryImportGenerator.dll
  System.Private.CoreLib -> C:\dev\dotnet\runtime4\artifacts\bin\System.Private.CoreLib\ref\Checked\net9.0\System.Private.CoreLib.dll
CSC : error CS8032: An instance of analyzer ILLink.RoslynAnalyzer.RequiresAssemblyFilesAnalyzer cannot be created from C:\dev\dotnet\runtime4\artifacts\bin\ILLink.RoslynAnalyzer\Release\netstandard2.0\ILLink.RoslynAnal
yzer.dll : Could not load file or assembly 'System.Collections.Immutable, Version=9.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'. The system cannot find the file specified.. [C:\dev\dotnet\runtime4\src\core
clr\System.Private.CoreLib\System.Private.CoreLib.csproj]
CSC : error CS8032: An instance of analyzer ILLink.RoslynAnalyzer.RequiresDynamicCodeAnalyzer cannot be created from C:\dev\dotnet\runtime4\artifacts\bin\ILLink.RoslynAnalyzer\Release\netstandard2.0\ILLink.RoslynAnalyz
er.dll : Could not load file or assembly 'System.Collections.Immutable, Version=9.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'. The system cannot find the file specified.. [C:\dev\dotnet\runtime4\src\corecl
r\System.Private.CoreLib\System.Private.CoreLib.csproj]
CSC : error CS8032: An instance of analyzer ILLink.RoslynAnalyzer.COMAnalyzer cannot be created from C:\dev\dotnet\runtime4\artifacts\bin\ILLink.RoslynAnalyzer\Release\netstandard2.0\ILLink.RoslynAnalyzer.dll : Could n
ot load file or assembly 'System.Collections.Immutable, Version=9.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'. The system cannot find the file specified.. [C:\dev\dotnet\runtime4\src\coreclr\System.Private
.CoreLib\System.Private.CoreLib.csproj]
CSC : error CS8032: An instance of analyzer ILLink.RoslynAnalyzer.DynamicallyAccessedMembersAnalyzer cannot be created from C:\dev\dotnet\runtime4\artifacts\bin\ILLink.RoslynAnalyzer\Release\netstandard2.0\ILLink.Rosly
nAnalyzer.dll : Could not load file or assembly 'System.Collections.Immutable, Version=9.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'. The system cannot find the file specified.. [C:\dev\dotnet\runtime4\src
\coreclr\System.Private.CoreLib\System.Private.CoreLib.csproj]
CSC : error CS8032: An instance of analyzer ILLink.RoslynAnalyzer.RequiresUnreferencedCodeAnalyzer cannot be created from C:\dev\dotnet\runtime4\artifacts\bin\ILLink.RoslynAnalyzer\Release\netstandard2.0\ILLink.RoslynA
nalyzer.dll : Could not load file or assembly 'System.Collections.Immutable, Version=9.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'. The system cannot find the file specified.. [C:\dev\dotnet\runtime4\src\c
oreclr\System.Private.CoreLib\System.Private.CoreLib.csproj]
```
Any idea why?

AlexRadch commented:
@sbomer I can not build runtime after this PR #106321

ViktorHofer commented on src/tools/illink/src/ILLink.RoslynAnalyzer/ILLink.RoslynAnalyzer.csproj:
```diff
@@ -28,6 +31,7 @@
       <PrivateAssets>all</PrivateAssets>
       <ExcludeAssets>contentfiles</ExcludeAssets> <!-- We include our own copy of the ClosedAttribute to work in source build -->
     </PackageReference>
+    <PackageReference Include="System.Reflection.Metadata" Version="$(SystemReflectionMetadataVersion)" />
```
This change seems to break the repo build. See https://github.com/dotnet/runtime/issues/106321

ericstj replied:
Yeah, an analyzer should only be using the version of SRM that roslyn provides for it.

It could be possible for an analyzer to carry a newer version of SRM, but it would need to be copied/deployed with the analyzer - and we'd need to make sure to never try to exchange SRM types with the compiler (I don't *think* any SRM types are part of roslyn's public API but I can't be certain).
